build_runner/lib/src/generate/build_impl.dart

```diff
@@ -189,9 +189,13 @@ class BuildImpl {
   }
 
   Set<AssetId> _inputsForPhase(int phaseNumber) => _assetGraph.allNodes
-      .where((n) =>
-          n is! GeneratedAssetNode ||
-          (n as GeneratedAssetNode).phaseNumber < phaseNumber)
+      .where((n) {
+        if (n is GeneratedAssetNode) {
+          return n.wasOutput == true && n.phaseNumber < phaseNumber;
+        } else {
+          return true;
```

Comment:
FWIW I think it's more readable in the old style rather than a boolean return inside a conditional. The repeated `as GeneratedAssetNode` does get noisy so I'm fine with this, but eventually we'll be able to drop the casts if the enhanced type promotion stuff goes through.